# HG changeset patch
# User Rik <rik@octave.org>
# Date 1487567403 28800
#      Sun Feb 19 21:10:03 2017 -0800
# Node ID 8fd8ed1e3aa2472b20e5c47dcbee662658d49457
# Parent  75515a7b1b4b474a1594ddb77576c025fb3f4e04
# Parent  9b0a2bc07545c75696d52c48a3d65f2d83767c1f
maint: merge of stable to default.

diff --git a/doc/interpreter/genpropdoc.m b/doc/interpreter/genpropdoc.m
--- a/doc/interpreter/genpropdoc.m
+++ b/doc/interpreter/genpropdoc.m
@@ -121,36 +121,37 @@ function s = getdoc (objname, field, bas
   ## in the list of each graphics object.  If a given graphics object
   ## interprets the property differently than others, then the doc will have
   ## to be rewritten for this object.
   if (strcmp (objname, "base"))
     switch (field)
       case "beingdeleted"
       case "busyaction"
       case "buttondownfcn"
+        s.doc = "__fcnmsg__";
         s.valid = valid_fcn;
 
       case "children"
         s.doc = "Graphics handles of the __objname__'s children.";
         s.valid = "vector of graphics handles";
 
       case "clipping"
         s.doc = "If __prop__ is @qcode{\"on\"}, the __objname__ is \
 clipped in its parent axes limits.";
 
       case "createfcn"
         s.doc = "Callback function executed immediately after __objname__ \
 has been created.  Function is set by using default property on root object, \
 e.g., @code{set (0, \"default__objname__createfcn\", \
-'disp (\"__objname__ created!\")')}.";
+'disp (\"__objname__ created!\")')}.\n\n__fcnmsg__";
         s.valid = valid_fcn;
 
       case "deletefcn"
         s.doc = "Callback function executed immediately before __objname__ \
-is deleted.";
+is deleted.\n\n__fcnmsg__";
         s.valid = valid_fcn;
 
       case "handlevisibility"
         s.doc = "If __prop__ is @qcode{\"off\"}, the __objname__'s \
 handle is not visible in its parent's \"children\" property.";
 
       case "hittest"
       case "interruptible"
@@ -285,17 +286,17 @@ are visible in their parents' children l
         s.doc = doc_unused;
 
       ## Specific properties
       case "alphamap"
         s.doc = sprintf (doc_notimpl, "Transparency");
 
       case "closerequestfcn"
         s.doc = "Function that is executed when a figure is deleted.  \
-@xref{XREFclosereq, , closereq function}."
+@xref{XREFclosereq, , closereq function}.\n\n__fcnmsg__";
         s.valid = valid_fcn;
 
       case "color"
         s.doc = "Color of the figure background.  \
 @xref{Colors, , colorspec}.";
         s.valid = valid_color;
 
       case "colormap"
@@ -341,33 +342,37 @@ only if the figure's callback @code{wind
 
       case "integerhandle"
         s.doc = "Assign the next lowest unused integer as the Figure number.";
 
       case "inverthardcopy"
         s.doc = "Replace the figure and axes background color with white when printing.";
 
       case "keypressfcn"
+        s.doc = "Callback function executed when a keystroke event \
+happens while the figure has focus. The actual key that was pressed \
+can be retrieved using the second argument 'evt' of the function.  __fcnmsg__";
         s.valid = valid_fcn;
 
       case "keyreleasefcn"
         s.doc = "With @code{keypressfcn}, the keyboard callback functions.  \
 These callback functions are called when a key is pressed/released \
 respectively.  The functions are called with two input arguments.  The first \
 argument holds the handle of the calling figure.  The second argument holds \
 an event structure which has the following members:\n\
 @table @code\n\
 @item Character:\n\
 The ASCII value of the key\n\
 @item Key:\n\
 Lowercase value of the key\n\
 @item Modifier:\n\
 A cell array containing strings representing the modifiers pressed with the \
 key.\n\
-@end table";
+@end table\
+\n\n__fcnmsg__";
         s.valid = valid_fcn;
 
       case "menubar"
         s.doc = "Control the display of the figure menu bar at the top \
 of the figure.";
 
       case "name"
         s.doc = "Name to be displayed in the figure title bar.  The name is \
@@ -468,17 +473,18 @@ and width and height of the figure.  \
       case "resizefcn"
         s.doc = "__prop__ is deprecated.  Use @code{sizechangedfcn} instead.";
         s.valid = valid_fcn;
 
       case "selectiontype"
         ## FIXME: docstring explaining what "{normal}|open|alt|extend" mean.
 
       case "sizechangedfcn"
-        s.doc = "Callback triggered when the figure window size is changed.";
+        s.doc = "Callback triggered when the figure window size is changed.\
+\n\n__fcnmsg__";
         s.valid = valid_fcn;
 
       case "toolbar"
         s.doc = "Control the display of the toolbar (along the bottom of the \
 menubar) and the status bar.  When set to @qcode{\"auto\"}, the display is based on the value of the @code{menubar} property.";
 
       case "units"
         s.doc = "The unit used to compute the @code{position} and \
@@ -494,30 +500,32 @@ menubar) and the status bar.  When set t
 @w{windowbuttonupfcn property}}.";
         s.valid = valid_fcn;
 
       case "windowbuttonupfcn"
         s.doc = "With @code{windowbuttondownfcn} and \
 @code{windowbuttonmotionfcn}, the mouse callback functions.  These \
 callback functions are called when a mouse button is pressed, dragged, or \
 released respectively.  When these callback functions are executed, the \
-@code{currentpoint} property holds the current coordinates of the cursor.";
+@code{currentpoint} property holds the current coordinates of the cursor.\
+\n\n__fcnmsg__";
         s.valid = valid_fcn;
 
       case "windowkeypressfcn"
         s.doc = "Function that is executed when a key is pressed and \
-the figure has focus.";
+the figure has focus.\n\n__fcnmsg__";
         s.valid = valid_fcn;
 
       case "windowkeyreleasefcn"
         s.doc = "Function that is executed when a key is released and \
-the figure has focus.";
+the figure has focus.\n\n__fcnmsg__";
         s.valid = valid_fcn;
 
       case "windowscrollwheelfcn"
+        s.doc = doc_unused;
         s.valid = valid_fcn;
 
       case "windowstyle"
         s.doc = "The window style of a figure.  One of the following values:\n\
 @table @code\n\
 @item normal\n\
 Set the window style as non modal.\n\
 @item modal\n\
@@ -1547,16 +1555,19 @@ endfunction
 function strout = expand_doc (strin, field, objname)
   strout = strrep (strin, "__objname__", objname);
   strout = strrep (strout, "__prop__", ["@code{" field "}"]);
 
   modemsg = "Setting @code{%s} also forces the @code{%smode} \
 property to be set to @qcode{\"manual\"}";
   modemsg = sprintf (modemsg, field, field);
   strout = strrep (strout, "__modemsg__", modemsg);
+  fcnmsg = "For information on how to write graphics listener \
+functions see @ref{Callbacks, , @w{Callbacks section}}.";
+  strout = strrep (strout, "__fcnmsg__", fcnmsg);
 endfunction
 
 function s = getstructure (objname, base = [])
   hf = [];
   if (! strcmp (objname, "root"))
     ## Use an improbable number to avoid ishandle to return true for 1
     hf = figure (2265465, "visible", "off");
   endif
diff --git a/doc/interpreter/plot.txi b/doc/interpreter/plot.txi
--- a/doc/interpreter/plot.txi
+++ b/doc/interpreter/plot.txi
@@ -1771,69 +1771,88 @@ and will return the color, line, and mar
 @cindex callbacks
 
 Callback functions can be associated with graphics objects and triggered
 after certain events occur.  The basic structure of all callback function
 is
 
 @example
 @group
-function mycallback (src, data)
+function mycallback (hsrc, evt)
 @dots{}
 endfunction
 @end group
 @end example
 
 @noindent
-where @code{src} gives a handle to the source of the callback, and
-@code{code} gives some event specific data.  This can then be associated
+where @code{hsrc} is a handle to the source of the callback, and @code{evt} 
+gives some event specific data.  
+
+The function can be provided as a function handle to a plain Octave function, 
+as an anonymous function or as a string representing an Octvae command. The 
+latter syntax is not recommended since syntax errors will only occur when the 
+string is evaluated.  
+@xref{Function Handles Anonymous Functions Inline Functions, , Function Handles section}.
+
+This can then be associated
 with an object either at the objects creation or later with the
 @code{set} function.  For example,
 
 @example
-plot (x, "DeleteFcn", @@(s, e) disp ("Window Deleted"))
+plot (x, "DeleteFcn", @@(h, e) disp ("Window Deleted"))
 @end example
 
 @noindent
 where at the moment that the plot is deleted, the message "Window
 Deleted" will be displayed.
 
 Additional user arguments can be passed to callback functions, and will
 be passed after the 2 default arguments.  For example:
 
 @example
 @group
 plot (x, "DeleteFcn", @{@@mycallback, "1"@})
 @dots{}
-function mycallback (src, data, a1)
+function mycallback (h, e, a1)
   fprintf ("Closing plot %d\n", a1);
 endfunction
 @end group
 @end example
 
 The basic callback functions that are available for all graphics objects
 are
 
 @itemize @bullet
-@item CreateFcn
-This is the callback that is called at the moment of the objects
-creation.  It is not called if the object is altered in any way, and so
-it only makes sense to define this callback in the function call that
-defines the object.  Callbacks that are added to @code{CreateFcn} later with
-the @code{set} function will never be executed.
-
-@item DeleteFcn
-This is the callback that is called at the moment an object is deleted.
-
-@item ButtonDownFcn
-This is the callback that is called if a mouse button is pressed while
-the pointer is over this object.  Note, that the gnuplot interface does
-not respect this callback.
+@item CreateFcn: 
+called at the moment of the objects creation.  It is not called if the object is altered in any way, and so it only makes sense to define this callback in the function call that defines the object.  Callbacks that are added to @code{CreateFcn} later with the @code{set} function will never be executed.
+
+@item DeleteFcn: 
+called at the moment an object is deleted.
+
+@item ButtonDownFcn: 
+called if a mouse button is pressed while the pointer is over this object.  Note, that the gnuplot interface does not respect this callback.
 @end itemize
 
+@strong{Caution:} the second @code{evt} argument in callback functions is only 
+loosely implemented in the Qt graphics toolkit:
+
+@itemize @bullet
+@item Mouse click events:
+@code{evt} is a class @code{double} value, 1 for left, 2 for middle and 3 
+for right click.
+
+@item Key press events:
+@code{evt} is a structure with fields @code{Key} (string), @code{Character} 
+(string) and @code{Modifier} (cell array of strings).
+
+@item Other events:
+@code{evt} is a class @code{double} empty matrix.
+@end itemize
+
+
 The object and figure that the event occurred in that resulted in the
 callback being called can be found with the @code{gcbo} and @code{gcbf}
 functions.
 
 @DOCSTRING(gcbo)
 
 @DOCSTRING(gcbf)
 
diff --git a/liboctave/system/file-ops.cc b/liboctave/system/file-ops.cc
--- a/liboctave/system/file-ops.cc
+++ b/liboctave/system/file-ops.cc
@@ -222,17 +222,17 @@ namespace octave
     // tilde.
 
     static std::string
     tilde_expand_word (const std::string& filename)
     {
       size_t f_len = filename.length ();
 
       if (f_len == 0 || filename[0] != '~')
-        return filename;
+        return std::string (filename);
 
       // A leading '~/' or a bare '~' is *always* translated to the value
       // of $HOME or the home directory of the current user, regardless of
       // any preexpansion hook.
 
       if (f_len == 1 || octave::sys::file_ops::is_dir_sep (filename[1]))
         return octave::sys::env::get_home_directory () + filename.substr (1);
 
@@ -296,17 +296,17 @@ namespace octave
 
     // If NAME has a leading ~ or ~user, Unix-style, expand it to the
     // user's home directory.  If no ~, or no <pwd.h>, just return NAME.
 
     std::string
     octave::sys::file_ops::tilde_expand (const std::string& name)
     {
       if (name.find ('~') == std::string::npos)
-        return name;
+        return std::string (name);
       else
         {
           std::string result;
 
           size_t name_len = name.length ();
 
           // Scan through S expanding tildes as we come to them.
 
diff --git a/liboctave/system/file-stat.cc b/liboctave/system/file-stat.cc
--- a/liboctave/system/file-stat.cc
+++ b/liboctave/system/file-stat.cc
@@ -162,16 +162,25 @@ namespace octave
     {
       file_stat fs (file);
 
       return fs ? fs.is_newer (time) : -1;
     }
 
     // Private stuff:
 
+    file_stat::file_stat (const std::string& n, bool fl)
+        : base_file_stat (), file_name (n), follow_links (fl)
+      {
+        if (! file_name.empty ())
+          update_internal ();
+      }
+
+    inline file_stat::~file_stat () { }
+
     void
     file_stat::update_internal (bool force)
     {
       if (! initialized || force)
         {
           initialized = false;
           fail = false;
 
diff --git a/liboctave/system/file-stat.h b/liboctave/system/file-stat.h
--- a/liboctave/system/file-stat.h
+++ b/liboctave/system/file-stat.h
@@ -206,22 +206,19 @@ namespace octave
     };
 
     class
     OCTAVE_API
     file_stat : public base_file_stat
     {
     public:
 
-      file_stat (const std::string& n = "", bool fl = true)
-        : base_file_stat (), file_name (n), follow_links (fl)
-      {
-        if (! file_name.empty ())
-          update_internal ();
-      }
+      // This constructor must remain defined in the cpp file rather than in
+      // the header file (bug #50234).
+      file_stat (const std::string& n = "", bool fl = true);
 
       file_stat (const file_stat& fs)
         : base_file_stat (fs), file_name (fs.file_name),
           follow_links (fs.follow_links) { }
 
       file_stat& operator = (const file_stat& fs)
       {
         if (this != &fs)
@@ -230,17 +227,19 @@ namespace octave
 
             file_name = fs.file_name;
             follow_links = fs.follow_links;
           }
 
         return *this;
       }
 
-      ~file_stat (void) = default;
+      // This destructor must remain as an empty destructor defined in the
+      // cpp file rather than in the header file (bug #50234).
+      ~file_stat (void);
 
       void get_stats (bool force = false)
       {
         if (! initialized || force)
           update_internal (force);
       }
 
       void get_stats (const std::string& n, bool force = false)
